### ISSUE TITLE
Split unit `impl` classes into separate files

### ISSUE THREAD

**Issue by younies** [icu-units/icu#44]
right now, `MeasureUnitImpl` and `SingleUnitImpl` headers in the same file and the implementation in two different file (measunit.cpp and measunit_extra.cpp)

I think it is better to have a file (.h and .cpp) for each one of them

For example:
   singleunitimpl.h and singleunitimpl.cpp
   measureunitimpl.h and measureunitimpl.cpp

**Comment by sffc** [comment]
The .cpp file has the definitions for the two classes, but it also contains data loading that's common between them, and it also includes a few methods for MeasureUnit. That being said, it's fine with me if you want to split them.

**Comment by younies** [comment]
Okay, I will try to do it through the current milestone

**Comment by hugovdm** [comment]
A quick word on priorities (talking to myself as much as to you):

Be sure to put priority on existing milestone goals: I'm also very prone to distraction by other tasks (hypocrisy? ;-) ), and some make sense (and some help me build momentum even if not a milestone), so I need to make sure I do give most time and energy to primary goals.

Regarding consistency between .h and .cpp files: I also like that. It's not possible to do it consistently within ICU though. For just one example, testlog.h defines IcuTestErrorCode. The implementation lives in tstdtmod.cpp. For sanity in C++, I'm trying to learn to use my tools well. I'm currently relying on VSCode's "go to definition" via "F12" - it seeks references when I'm already on the definition. I usually find what I'm looking for without too much trouble. And I'm sure I can improve my use of the tools further.

In conclusion: do be careful with "busywork" in cases where it may take much more time than you think (I make that mistake too often) - and do be aware that complete consistency between .h and .cpp files is not an attainable goal within ICU.

**Comment by sffc** [comment]
> do be aware that complete consistency between .h and .cpp files is not an attainable goal within ICU.

I would go further than that and say that .h and .cpp files serve different goals, and full parity between them is a non-goal of ICU.

.h files represent logical groupings of classes based on visibility (internal or external).  It is common to have a public header file and a corresponding internal header file, like unicode/plurrule.h and plurrule_impl.h, or unicode/measunit.h and measunit_impl.h.  However, the definitions for those classes might share the same .cpp file.

.cpp files represent code modules for the purposes of dependency resolution.  For example, the identifier parsing functions for measurement units are declared in unicode/measunit.h, but defined in measunit_extra.cpp, because not everyone using MeasureUnit needs the identifier parsing functionality.

Since MeasureUnitImpl and SingleUnitImpl are strongly coupled, there's no benefit in terms of dependency resolution for splitting them into separate cpp files.  The only benefit would be for code readability, which is something in the eyes of the observer.  I personally don't think there's much benefit for code readability, and personally I think it might slightly detract from it, since there are some shared functions in measunit_extra.cpp, for things like data loading.  However, since Younies thinks that splitting this file would improve code readability, that's strong enough signal that maybe it's a good idea.
